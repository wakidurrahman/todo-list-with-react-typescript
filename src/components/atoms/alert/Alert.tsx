/**
 * Alert Component
 *
 * A reusable alert component that uses Bootstrap classes for styling
 * and includes accessibility features.
 *
 * Performance Optimizations:
 *
 * 1. useCallback for handleClose:
 * - Memoizes the close handler function to maintain referential equality between renders
 * - Prevents unnecessary re-renders of child components that receive this as a prop
 * - Only recreates if onClose prop changes
 * - Important since this is passed to Button component and used in useEffect
 *
 * 2. useMemo for alertClassName:
 * - Memoizes the computed className string to avoid recalculating on every render
 * - Only recomputes when variant or dismissible props change
 * - While string concatenation is not expensive, memoization helps when component re-renders frequently
 * - Follows best practice of caching computed values used in render
 *
 * 3. useState for isVisible:
 * - Manages internal visibility state separate from props
 * - Allows for smooth transitions and animations
 * - Enables controlled dismissal behavior
 *
 * 4. useEffect for timeout:
 * - Handles auto-dismiss functionality
 * - Properly cleans up timeout to prevent memory leaks
 * - Only runs when relevant dependencies change
 *
 * @component
 * @example
 * ```tsx
 * <Alert
 *   message="Success!"
 *   variant="success"
 *   dismissible={true}
 *   timeout={3000}
 *   onClose={() => console.log('Alert closed')}
 * />
 * ```
 */

import { useCallback, useEffect, useMemo, useState } from 'react';

import Button from '@/components/atoms/button/Button';
import { AlertVariant } from '@/types/error.types';

type AlertProps = {
  /** Whether the alert can be dismissed */
  dismissible?: boolean;
  /** Content to display in the alert */
  message: React.ReactNode;
  /** Whether the alert is visible */
  show?: boolean;
  /** Auto-dismiss timeout in milliseconds */
  timeout?: number;
  /** Bootstrap alert variant */
  variant?: AlertVariant;
  /** Callback when alert is closed */
  onClose?: () => void;
};

const Alert = ({
  dismissible = true,
  message,
  show = true,
  timeout,
  variant = 'info',
  onClose,
}: AlertProps) => {
  const [isVisible, setIsVisible] = useState<boolean>(show);

  // Memoize handleClose to prevent recreation on each render
  const handleClose = useCallback(() => {
    setIsVisible(false);
    onClose?.();
  }, [onClose]);

  useEffect(() => {
    setIsVisible(show);

    if (timeout && show) {
      const timer = setTimeout(handleClose, timeout);
      return () => clearTimeout(timer);
    }
    return () => {}; // Add empty cleanup function for consistent return
  }, [timeout, show, handleClose]);

  const alertClassName = useMemo(
    () =>
      `alert alert-${variant} ${
        dismissible ? 'alert-dismissible fade show' : ''
      }`,
    [variant, dismissible]
  );

  // Early return for hidden alert
  if (!isVisible) return null;

  return (
    <div
      className={alertClassName}
      role="alert"
      aria-live="polite"
      data-testid="alert"
    >
      {message}
      {dismissible && (
        <Button
          variant="secondary"
          className="btn-close"
          aria-label="Close alert"
          onClick={handleClose}
          data-testid="alert-close-button"
        />
      )}
    </div>
  );
};

export default Alert;
