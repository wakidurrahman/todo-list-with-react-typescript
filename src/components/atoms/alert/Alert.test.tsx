import { act } from 'react';

import '@testing-library/jest-dom';
import { fireEvent, render, screen } from '@testing-library/react';
import { create } from 'react-test-renderer';

import Alert from './Alert';

/**
 * Alert Component Test Suite
 *
 * A comprehensive test suite covering all major functionality:
 *
 * 1. Basic Rendering
 *    - Default props rendering
 *    - Custom variant rendering
 *
 * 2. Dismissible Functionality
 *    - Close button presence/absence
 *    - Close button click handling
 *
 * 3. Show/Hide Behavior
 *    - Visibility toggle with show prop
 *
 * 4. Timeout Functionality
 *    - Auto-dismiss after timeout
 *    - Cleanup on unmount
 *
 * 5. Accessibility Features
 *    - ARIA attributes
 *    - Role attributes
 *
 * 6. Message Rendering
 *    - Text messages
 *    - Complex ReactNode messages
 */

describe('Alert Component', () => {
  beforeEach(() => {
    jest.useFakeTimers();
  });

  afterEach(() => {
    jest.restoreAllMocks();
  });

  // Basic rendering tests
  it('renders with default props', () => {
    render(<Alert message="Test message" />);

    const alert = screen.getByRole('alert');
    expect(alert).toBeInTheDocument();
    expect(alert).toHaveClass('alert', 'alert-info', 'alert-dismissible');
    expect(screen.getByText('Test message')).toBeInTheDocument();
  });

  it('renders with custom variant', () => {
    render(<Alert message="Success message" variant="success" />);

    const alert = screen.getByRole('alert');
    expect(alert).toHaveClass('alert-success');
  });

  // Dismissible functionality tests
  it('renders without close button when dismissible is false', () => {
    render(<Alert message="Non-dismissible alert" dismissible={false} />);

    expect(screen.queryByLabelText('Close alert')).not.toBeInTheDocument();
  });

  it('calls onClose when close button is clicked', () => {
    const handleClose = jest.fn();
    render(<Alert message="Closeable alert" onClose={handleClose} />);

    const closeButton = screen.getByLabelText('Close alert');
    fireEvent.click(closeButton);

    expect(handleClose).toHaveBeenCalledTimes(1);
  });

  // Show/hide behavior tests
  it('hides when show prop is false', () => {
    const { rerender } = render(<Alert message="Hidden alert" show={false} />);

    expect(screen.queryByRole('alert')).not.toBeInTheDocument();

    rerender(<Alert message="Visible alert" show={true} />);
    expect(screen.getByRole('alert')).toBeInTheDocument();
  });

  // Timeout tests
  it('auto-dismisses after timeout', () => {
    const handleClose = jest.fn();
    render(
      <Alert
        message="Auto-dismiss alert"
        timeout={3000}
        onClose={handleClose}
      />
    );

    expect(screen.getByRole('alert')).toBeInTheDocument();

    act(() => {
      jest.advanceTimersByTime(3000);
    });

    expect(handleClose).toHaveBeenCalledTimes(1);
  });

  // Accessibility tests
  it('has correct accessibility attributes', () => {
    render(<Alert message="Accessible alert" />);

    const alert = screen.getByRole('alert');
    expect(alert).toHaveAttribute('aria-live', 'polite');

    const closeButton = screen.getByLabelText('Close alert');
    expect(closeButton).toHaveAttribute('aria-label', 'Close alert');
  });

  // Message rendering tests
  it('renders ReactNode message correctly', () => {
    const complexMessage = (
      <div>
        <h4>Alert Title</h4>
        <p>Alert content</p>
      </div>
    );

    render(<Alert message={complexMessage} />);

    expect(screen.getByText('Alert Title')).toBeInTheDocument();
    expect(screen.getByText('Alert content')).toBeInTheDocument();
  });

  // Cleanup test
  it('cleans up timeout on unmount', () => {
    const handleClose = jest.fn();
    const { unmount } = render(
      <Alert message="Cleanup test" timeout={3000} onClose={handleClose} />
    );

    unmount();

    act(() => {
      jest.advanceTimersByTime(3000);
    });

    expect(handleClose).not.toHaveBeenCalled();
  });

  // Snapshot tests
  describe('Snapshots', () => {
    it('matches default alert snapshot', () => {
      const tree = create(<Alert message="Default alert" />).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('matches success variant snapshot', () => {
      const tree = create(
        <Alert message="Success alert" variant="success" />
      ).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('matches non-dismissible alert snapshot', () => {
      const tree = create(
        <Alert message="Non-dismissible alert" dismissible={false} />
      ).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('matches complex message snapshot', () => {
      const complexMessage = (
        <div>
          <h4>Alert Title</h4>
          <p>Alert content</p>
        </div>
      );
      const tree = create(<Alert message={complexMessage} />).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
});
