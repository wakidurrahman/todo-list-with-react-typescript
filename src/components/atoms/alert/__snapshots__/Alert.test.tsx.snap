// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Alert Component Snapshots matches complex message snapshot 1`] = `
<div
  aria-live="polite"
  className="alert alert-info alert-dismissible fade show"
  data-testid="alert"
  role="alert"
>
  <div>
    <h4>
      Alert Title
    </h4>
    <p>
      Alert content
    </p>
  </div>
  <button
    aria-label="Close alert"
    className="btn btn-secondary btn-md btn-close"
    data-testid="alert-close-button"
    disabled={false}
    onClick={[Function]}
    type="button"
  />
</div>
`;

exports[`Alert Component Snapshots matches default alert snapshot 1`] = `
<div
  aria-live="polite"
  className="alert alert-info alert-dismissible fade show"
  data-testid="alert"
  role="alert"
>
  Default alert
  <button
    aria-label="Close alert"
    className="btn btn-secondary btn-md btn-close"
    data-testid="alert-close-button"
    disabled={false}
    onClick={[Function]}
    type="button"
  />
</div>
`;

exports[`Alert Component Snapshots matches non-dismissible alert snapshot 1`] = `
<div
  aria-live="polite"
  className="alert alert-info "
  data-testid="alert"
  role="alert"
>
  Non-dismissible alert
</div>
`;

exports[`Alert Component Snapshots matches success variant snapshot 1`] = `
<div
  aria-live="polite"
  className="alert alert-success alert-dismissible fade show"
  data-testid="alert"
  role="alert"
>
  Success alert
  <button
    aria-label="Close alert"
    className="btn btn-secondary btn-md btn-close"
    data-testid="alert-close-button"
    disabled={false}
    onClick={[Function]}
    type="button"
  />
</div>
`;
